spack docs draft 1 (#273)
ResearchComputing/Documentation: docs/software/spack.md

```
[johndoe@c3cpu-c11-u17-2 ~]$ spack install --add gcc@13.1.0
```

Once the installation is complete, you can make it available to Spack with `spack compiler add`:

```
[johndoe@c3cpu-c11-u17-2 ~]$ spack compiler add <path to compiler>
```

>Note that, in this case, you can locate the path of the compiler with `spack location -i gcc@13.1.0`.

Once the compiler is added, you can install any subsequent packages using the compiler you've installed with the `%` operator. For example:

```
[johndoe@c3cpu-c11-u17-2 ~]$ spack install --add fastqc%gcc@13.1.0
```

### Key Spack Terms: 

| Term | Definition |
|---------|----------|
| Spec | Description of the build requirements for an installation. This includes the software, version, compiler, and other variants. |
| Variant | An option or feature that can be toggled for a specific package. These usually correspond to a configuration or compilation setting. |
| Repo | A central location which stores packages. Repo locations can be configured. |
| Environment | A collection of packages and dependencies that are available to a user. In Spack, environments do not interact. |

### Basic Spack commands:

| Command | Function |

> |---------|----------|

Reviewer: add:
spack env list (list all your environments)
spack remove <env> (remove an environment) 
spack uninstall <packagename> remove a package - though this can get complicated so maybe link these uninstall docs - https://spack-tutorial.readthedocs.io/en/latest/tutorial_environments.html#uninstalling-packages
spack env status <check which env you're in>

| `spack compilers` | Lists available compilers |
| `spack spec <software>` | Shows all dependencies and items that would be installed if `<software>` is installed > |

### Troubleshooting Spack

If an installation fails, there are several ways to troubleshoot the failure. Common issues with installations include: 

* __Building with the wrong compiler:__ Double check if the compiler you are using is compatible with the software you are attempting to build. 
* __Building with an incompatible variant set:__ Confirm that the configuration settings for the software are compatible and as expected. If not, try enabling or disabling variants that may be causing problems. 
* __Building an unexpected version:__ Make sure you are building the intended version of a software. If a new version is failing, try insalling an older version. 
* __Building with an unexpected version of a dependency:__ If there are issues installing a dependency, you can modify the specs of a dependency.

Additionally, spack will oftentimes output troubleshooting suggestions. To increase the verbosity of `spack install`, instead use `spack -dv install`. 

> 

Reviewer: Maybe include 'fill out our help form if you have any questions (with link to help form)'

# Spack

*Currently, Spack is only available on Alpine.* [Spack](https://spack.readthedocs.io/en/latest/) is a package manager designed for installing multiple configurations. Spack is designed for Supercomputing clusters, making it easy to install and configure software that may not currently be available in the Alpine software stack. The following documentation demonstrates how to activate & use the CURC Spack module and utilize your own Spack environments to install packages and compilers.

## Using Spack on CURC
You can follow these steps from a Research Computing terminal session via a compile job (`acompile`) or within a batch or interactive job.

### Activate and Use the CURC Spack Module

Run the following command to load the Spack module:  

```

> [johndoe@c3cpu-c11-u17-2 ~]$ module load spack/0.20.1

Reviewer 2: ```suggestion
[johndoe@c3cpu-c11-u17-2 ~]$ module load spack
```

I recommend the following. I think we should stop using specific versions in our documentation, unless we are demonstrating something where is makes sense. By adding versions, it makes the documentation harder to maintain.

```

You can confirm that spack has been loaded and find a list of useful spack commands by running the `spack` command: 

```

> [johndoe@c3cpu-c11-u17-2 ~]$ spack

Reviewer 2: ```suggestion
[johndoe@c3cpu-c11-u17-2 ~]$ spack help --all
```

If you want to provide all of the options for Spack, I would go with my suggestion.

```

or

```
[johndoe@c3cpu-c11-u17-2 ~]$ spack env deactivate
```

### Installing Software with Spack

Once you've created and activated an environment, you can begin installing software using Spack. To see a list of *all* of the software available, you can use the spack list command: 

```
[johndoe@c3cpu-c11-u17-2 ~]$ spack list
```

The above command lists all 7000+ packages available for installation with Spack. As a result, this command may take some time to execute.  To search for a specific software by name, you can use `spack list <software name>`. For example: 

```
[johndoe@c3cpu-c11-u17-2 ~]$ spack list fastqc
fastqc
==> 1 packages
```

>For additional information about a specific piece of software (available versions, variants, and dependencies), you can use the `spack info <software name>` command. Note that the 'preferred version' is the default version which will be installed if no alternative version is specified. 

Once you've confirmed that a piece of software is available, you can install it in the active environment using the `spack install --add` command:

```
[johndoe@c3cpu-c11-u17-2 ~]$ spack install --add fastqc
```

> >Note that the `--add` argument is required to add package specs to an environment. You can separately add specs without installing a package using the `spack add <software name>` command.

Reviewer 2: ```suggestion
>__Note:__
>- The `--add` argument is required to add package specs to an environment. You can separately add specs >without installing a package using the `spack add <software name>` command.
>- If no compiler is specified, then a default compiler will be installed and used within the environment.
```
I think my markdown above may not work, please double check it.

This will install the specified software and all dependencies into your environment.

You can also modify the installation commands to meet your installation needs. You can specify which version you'd like installed using the `@` operator:

```
[johndoe@c3cpu-c11-u17-2 ~]$ spack install --add fastqc@0.11.9
```

> 

Reviewer 2: ```suggestion
Once you have installed a package, you can view what packages are installed in your environment using the `spack find` command. For example, 
```code
[johndoe@c3cpu-c11-u17-2 ~]$ spack find
==> In environment my_test_environment
==> Root specs
fastqc@0.11.9

==> Installed packages
-- linux-rhel8-zen / gcc@8.5.0 ----------------------------------
berkeley-db@18.1.40  fastqc@0.11.9  ncurses@6.4        pkgconf@1.9.5
bzip2@1.0.8          gdbm@1.23      openjdk@11.0.17_8  readline@8.2
diffutils@3.9        libiconv@1.17  perl@5.36.0        zlib@1.2.13
==> 12 installed packages
```code
>Note that since we did not specify a compiler, the default compiler of `gcc@8.5.0` was used. Please also note that the default compiler may change. 

```

I am not exactly sure "```code" is the correct formatting, I had to do that because of the GitHub suggestions. Please double check the formatting.

The root of the Spack installation tree, by default, is `/projects/$USER/software/spack` (this can be changed in `~/config.yaml`). From there, executables can be found in subdirectories depending on the OS and compiler used to install the software. Software will be installed here regardless of whether the software was installed in an environment. 

Spack will always check the installation root prior to installing software, in an environment or otherwise. If the software is not found in the installation root, it will be installed there. If the software is found in the installation root, it will not be installed. When attempting to install in an environment, symbolic links to the installation root will be created in `/projects/$USER/spack/environments/<envname>/.spack-env/view`. This behavior means that once a piece of software is installed, it can be linked to any environment without having to perform the installation again. 

The default cache is located at `/scratch/alpine/$USER/spack/cache`. This directory will contain any previously-installed tarballs and repositories, as well as any miscellaneous cache items.  

### Installing and Using Compilers with Spack

> In addition to standard software packages, you can use Spack to install compilers which are not currently available in the Alpine software stack. For example, if you need the newest available version of `gcc`, you can install it into your environment as follows: 
> 
> ```
> [johndoe@c3cpu-c11-u17-2 ~]$ spack install --add gcc@13.1.0
> ```

Reviewer 2: ```suggestion
In addition to standard software packages, you can use Spack to install compilers which are not currently available in the Alpine software stack. It is highly recommended to __NOT__ install compilers directly within the environment. Instead, we suggest that you first install the compiler outside of the environment. If it is installed directly within an environment, then it will depend on the default compiler within the environment. The following is our recommended way to install a new compiler and use it as the compiler for your environment. 

1. Install the compiler outside of your environment (only needs to be done once):
```code
spack install gcc@13.1.0
```code

2. Obtain the location of the compiler install and save it into a local environmental variable:
```code
gcc_location=$(spack location -i gcc@13.1.0)
```code

3. Create and activate your environment:
```code
spack env create my_test_env
spack env activate my_test_env
```code

4. Remove any compilers that are in the env right now:
```code
spack compiler remove gcc -a
```code
>Note that you can also remove intel compilers using `spack compiler remove intel -a`

5. Add the compiler you installed outside of the environment and install the compiler into the environment:
```code
spack compiler add $gcc_location
spack install --add gcc@13.1.0
```code
>Note that the install of the compiler within the environment is necessary as it will install all dependencies needed for the compiler. 
```

Please check the formatting on the "```code" block, like in my previous comment.

> Once the installation is complete, you can make it available to Spack with `spack compiler add`:
> 
> ```
> [johndoe@c3cpu-c11-u17-2 ~]$ spack compiler add <path to compiler>
> ```
> 
> >Note that, in this case, you can locate the path of the compiler with `spack location -i gcc@13.1.0`.

Reviewer 2: ```suggestion
```

I believe you can remove this portion because I use and explain these commands in my suggested steps above.

| `spack env status` | Check which environment is currently activated |
| `spack remove <environment name>` | Remove an environment |
| `spacktivate <environment name>` | Activates a Spack environment |
| `despacktivate` | Deactivates a Spack environment |
| `spack find` | List and search packages installed in the active environment |
| `spack install --add <software>` | Adds specs for selected package and installs it in the active environment |
| `spack uninstall <software>` | Removes package from the activate environment. Refer to the [spack documentation](https://spack-tutorial.readthedocs.io/en/latest/tutorial_environments.html#uninstalling-packages) for additional information |
| `spack compilers` | Lists available compilers |
| `spack spec <software>` | Shows all dependencies and items that would be installed if `<software>` is installed > |

### Troubleshooting Spack

If an installation fails, there are several ways to troubleshoot the failure. Common issues with installations include: 

* __Building with the wrong compiler:__ Double check if the compiler you are using is compatible with the software you are attempting to build. 

> * __Building with an incompatible variant set:__ Confirm that the configuration settings for the software are compatible and as expected. If not, try enabling or disabling variants that may be causing problems. 

Reviewer 2: ```suggestion
* __Building with an incompatible variant set:__ Confirm that the configuration settings for the software are compatible and as expected. If not, try enabling or disabling variants that may be causing problems. 
    * One can easily see variants and possible conflicts by searching the package using [Spack's provided search engine](https://packages.spack.io/) and viewing the "Variants" and "Conflicts" section of the documentation. Additionally, one can view `package.py` in the "Build System" section to directly view the code that defines variants and conflicts.
```

* __Building an unexpected version:__ Make sure you are building the intended version of a software. If a new version is failing, try insalling an older version. 
* __Building with an unexpected version of a dependency:__ If there are issues installing a dependency, you can modify the specs of a dependency.

> Additionally, spack will oftentimes output troubleshooting suggestions. To increase the verbosity of `spack install`, instead use `spack -dv install`. 
> 
> Fill out our [help form](https://www.colorado.edu/rc/userservices/contact) if you need additional assistance. This section references Spack documentation written by [NERSC](https://docs.nersc.gov/development/build-tools/spack/).
> 

Reviewer 2: ```suggestion
Additionally, Spack will oftentimes output troubleshooting suggestions. To increase the verbosity of `spack install`, use `spack -dv install`. 

__Need further help?__ Fill out our [help form](https://www.colorado.edu/rc/userservices/contact) and we would be happy to help you with any further questions you may have. 

Suggestions provided in the `Troubleshooting Spack` subsection were based off of Spack documentation written by [NERSC](https://docs.nersc.gov/development/build-tools/spack/).
```
